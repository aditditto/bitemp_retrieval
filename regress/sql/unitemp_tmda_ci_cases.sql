DROP EXTENSION bitemp_retrieval;
CREATE EXTENSION bitemp_retrieval;
SET DATESTYLE TO 'ISO';
SET TIMEZONE TO 'Asia/Jakarta';


SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'empl',
	$$n text not null, 
        cid int,
        d text not null,
        p text not null,
        h int not null,
        s int not null
	$$,
   'cid');

-- r1
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Jan', 140, 'DB', 'P1', 2400, 1200$$,
    temporal_relationships.timeperiod('1970-01-01'::timestamptz, '1970-01-16'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r2
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Jan', 163, 'DB', 'P1', 600, 1500$$,
    temporal_relationships.timeperiod('1970-01-19'::timestamptz, '1970-01-22'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r3
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Ann', 141, 'DB', 'P2', 500, 700$$,
    temporal_relationships.timeperiod('1970-01-01'::timestamptz, '1970-01-06'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r4
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Ann', 150, 'DB', 'P1', 1000, 800$$,
    temporal_relationships.timeperiod('1970-01-06'::timestamptz, '1970-01-16'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r5
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Ann', 157, 'DB', 'P1', 600, 500$$,
    temporal_relationships.timeperiod('1970-01-13'::timestamptz, '1970-01-25'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r6
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Sue', 142, 'DB', 'P2', 400, 800$$,
    temporal_relationships.timeperiod('1970-01-01'::timestamptz, '1970-01-11'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r7
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Tom', 143, 'AI', 'P2', 1200, 2000$$,
    temporal_relationships.timeperiod('1970-01-04'::timestamptz, '1970-01-11'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

-- r8
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Tom', 153, 'AI', 'P1', 900, 1800$$,
    temporal_relationships.timeperiod('1970-01-13'::timestamptz, '1970-01-19'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT n, cid, d, p, h, s, effective FROM empl WHERE now() <@ asserted ORDER BY LOWER(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['avg'],
    ARRAY['h'],
    ARRAY['avg_h']
) AS (
    d text,
    avg_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['sum'],
    ARRAY['h'],
    ARRAY['sum_h']
) AS (
    d text,
    sum_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['max'],
    ARRAY['h'],
    ARRAY['max_h']
) AS (
    d text,
    max_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['min'],
    ARRAY['h'],
    ARRAY['min_h']
) AS (
    d text,
    min_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['count'],
    ARRAY['h'],
    ARRAY['count_h']
) AS (
    d text,
    count_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM bitemporal_internal.ll_register_temporal_attribute_property('public', 'empl', 'h', 'malleable');

SELECT * FROM bitemporal_internal.temporal_attribute_properties;

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['avg'],
    ARRAY['h'],
    ARRAY['avg_h']
) AS (
    d text,
    avg_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['sum'],
    ARRAY['h'],
    ARRAY['sum_h']
) AS (
    d text,
    sum_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['max'],
    ARRAY['h'],
    ARRAY['max_h']
) AS (
    d text,
    max_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['min'],
    ARRAY['h'],
    ARRAY['min_h']
) AS (
    d text,
    min_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['count'],
    ARRAY['h'],
    ARRAY['count_h']
) AS (
    d text,
    count_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

DELETE FROM bitemporal_internal.temporal_attribute_properties WHERE attr_name = 'h';
SELECT * FROM bitemporal_internal.temporal_attribute_properties;

SELECT * FROM bitemporal_internal.ll_register_temporal_attribute_property('public', 'empl', 'h', 'atomic');

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['avg'],
    ARRAY['h'],
    ARRAY['avg_h']
) AS (
    d text,
    avg_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['sum'],
    ARRAY['h'],
    ARRAY['sum_h']
) AS (
    d text,
    sum_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['max'],
    ARRAY['h'],
    ARRAY['max_h']
) AS (
    d text,
    max_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['min'],
    ARRAY['h'],
    ARRAY['min_h']
) AS (
    d text,
    min_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);

SELECT * FROM tmda_ci_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['count'],
    ARRAY['h'],
    ARRAY['count_h']
) AS (
    d text,
    count_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);