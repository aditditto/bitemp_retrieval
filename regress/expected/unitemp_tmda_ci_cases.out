DROP EXTENSION bitemp_retrieval;
CREATE EXTENSION bitemp_retrieval;
SET DATESTYLE TO 'ISO';
SET TIMEZONE TO 'Asia/Jakarta';
SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'empl',
	$$n text not null, 
        cid int,
        d text not null,
        p text not null,
        h int not null,
        s int not null
	$$,
   'cid');
 ll_create_bitemporal_table 
----------------------------
 t
(1 row)

-- r1
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Jan', 140, 'DB', 'P1', 2400, 1200$$,
    temporal_relationships.timeperiod('1970-01-01'::timestamptz, '1970-01-16'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r2
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Jan', 163, 'DB', 'P1', 600, 1500$$,
    temporal_relationships.timeperiod('1970-01-19'::timestamptz, '1970-01-22'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r3
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Ann', 141, 'DB', 'P2', 500, 700$$,
    temporal_relationships.timeperiod('1970-01-01'::timestamptz, '1970-01-06'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r4
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Ann', 150, 'DB', 'P1', 1000, 800$$,
    temporal_relationships.timeperiod('1970-01-06'::timestamptz, '1970-01-16'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r5
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Ann', 157, 'DB', 'P1', 600, 500$$,
    temporal_relationships.timeperiod('1970-01-13'::timestamptz, '1970-01-25'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r6
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Sue', 142, 'DB', 'P2', 400, 800$$,
    temporal_relationships.timeperiod('1970-01-01'::timestamptz, '1970-01-11'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r7
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Tom', 143, 'AI', 'P2', 1200, 2000$$,
    temporal_relationships.timeperiod('1970-01-04'::timestamptz, '1970-01-11'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

-- r8
SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.empl',
    $$n, cid, d, p, h, s$$,
    $$'Tom', 153, 'AI', 'P1', 900, 1800$$,
    temporal_relationships.timeperiod('1970-01-13'::timestamptz, '1970-01-19'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT n, cid, d, p, h, s, effective FROM empl WHERE now() <@ asserted ORDER BY LOWER(effective);
  n  | cid | d  | p  |  h   |  s   |                      effective                      
-----+-----+----+----+------+------+-----------------------------------------------------
 Sue | 142 | DB | P2 |  400 |  800 | ["1970-01-01 00:00:00+07","1970-01-11 00:00:00+07")
 Ann | 141 | DB | P2 |  500 |  700 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 Jan | 140 | DB | P1 | 2400 | 1200 | ["1970-01-01 00:00:00+07","1970-01-16 00:00:00+07")
 Tom | 143 | AI | P2 | 1200 | 2000 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 Ann | 150 | DB | P1 | 1000 |  800 | ["1970-01-06 00:00:00+07","1970-01-16 00:00:00+07")
 Tom | 153 | AI | P1 |  900 | 1800 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
 Ann | 157 | DB | P1 |  600 |  500 | ["1970-01-13 00:00:00+07","1970-01-25 00:00:00+07")
 Jan | 163 | DB | P1 |  600 | 1500 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
(8 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['avg'],
    ARRAY['h'],
    ARRAY['avg_h']
) AS (
    d text,
    avg_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  |      avg_h       |                      effective                      
----+------------------+-----------------------------------------------------
 DB |             1100 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB | 1266.66666666667 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |             1700 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB | 1333.33333333333 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |              600 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |              600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |              600 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |             1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |              900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['sum'],
    ARRAY['h'],
    ARRAY['sum_h']
) AS (
    d text,
    sum_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | sum_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |  3300 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |  3800 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |  3400 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |  4000 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   600 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |  1200 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   600 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['max'],
    ARRAY['h'],
    ARRAY['max_h']
) AS (
    d text,
    max_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | max_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |  2400 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |  2400 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |  2400 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |  2400 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   600 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   600 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['min'],
    ARRAY['h'],
    ARRAY['min_h']
) AS (
    d text,
    min_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | min_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   400 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   400 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |  1000 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |   600 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   600 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   600 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['count'],
    ARRAY['h'],
    ARRAY['count_h']
) AS (
    d text,
    count_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | count_h |                      effective                      
----+---------+-----------------------------------------------------
 DB |       3 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |       3 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |       2 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |       3 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |       1 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |       2 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |       1 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |       1 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |       1 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM bitemporal_internal.ll_register_temporal_attribute_property('public', 'empl', 'h', 'malleable');
 ll_register_temporal_attribute_property 
-----------------------------------------
                                       1
(1 row)

SELECT * FROM bitemporal_internal.temporal_attribute_properties;
 schema_name | table_name | attr_name | attr_property 
-------------+------------+-----------+---------------
 public      | empl       | h         | malleable
(1 row)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['avg'],
    ARRAY['h'],
    ARRAY['avg_h']
) AS (
    d text,
    avg_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | avg_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   500 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   500 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |   260 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |   310 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   150 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |   375 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   150 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['sum'],
    ARRAY['h'],
    ARRAY['sum_h']
) AS (
    d text,
    sum_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | sum_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |  1500 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |  1500 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |   520 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |   930 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   150 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |   750 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   150 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['max'],
    ARRAY['h'],
    ARRAY['max_h']
) AS (
    d text,
    max_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | max_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   800 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   800 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |   320 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |   480 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   150 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   150 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['min'],
    ARRAY['h'],
    ARRAY['min_h']
) AS (
    d text,
    min_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | min_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   200 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   200 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |   200 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |   150 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |   150 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |   150 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |   150 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['count'],
    ARRAY['h'],
    ARRAY['count_h']
) AS (
    d text,
    count_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | count_h |                      effective                      
----+---------+-----------------------------------------------------
 DB |       3 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |       3 | ["1970-01-06 00:00:00+07","1970-01-11 00:00:00+07")
 DB |       2 | ["1970-01-11 00:00:00+07","1970-01-13 00:00:00+07")
 DB |       3 | ["1970-01-13 00:00:00+07","1970-01-16 00:00:00+07")
 DB |       1 | ["1970-01-16 00:00:00+07","1970-01-19 00:00:00+07")
 DB |       2 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 DB |       1 | ["1970-01-22 00:00:00+07","1970-01-25 00:00:00+07")
 AI |       1 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |       1 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(9 rows)

DELETE FROM bitemporal_internal.temporal_attribute_properties WHERE attr_name = 'h';
SELECT * FROM bitemporal_internal.temporal_attribute_properties;
 schema_name | table_name | attr_name | attr_property 
-------------+------------+-----------+---------------
(0 rows)

SELECT * FROM bitemporal_internal.ll_register_temporal_attribute_property('public', 'empl', 'h', 'atomic');
 ll_register_temporal_attribute_property 
-----------------------------------------
                                       1
(1 row)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['avg'],
    ARRAY['h'],
    ARRAY['avg_h']
) AS (
    d text,
    avg_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | avg_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   500 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(4 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['sum'],
    ARRAY['h'],
    ARRAY['sum_h']
) AS (
    d text,
    sum_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | sum_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   500 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(4 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['max'],
    ARRAY['h'],
    ARRAY['max_h']
) AS (
    d text,
    max_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | max_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   500 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(4 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['min'],
    ARRAY['h'],
    ARRAY['min_h']
) AS (
    d text,
    min_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | min_h |                      effective                      
----+-------+-----------------------------------------------------
 DB |   500 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |   600 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 AI |  1200 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |   900 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(4 rows)

SELECT * FROM ita_now(
    'public',
    'empl',
    ARRAY['d'],
    ARRAY['count'],
    ARRAY['h'],
    ARRAY['count_h']
) AS (
    d text,
    count_h numeric,
    effective tstzrange
) ORDER BY d DESC, lower(effective);
 d  | count_h |                      effective                      
----+---------+-----------------------------------------------------
 DB |       1 | ["1970-01-01 00:00:00+07","1970-01-06 00:00:00+07")
 DB |       1 | ["1970-01-19 00:00:00+07","1970-01-22 00:00:00+07")
 AI |       1 | ["1970-01-04 00:00:00+07","1970-01-11 00:00:00+07")
 AI |       1 | ["1970-01-13 00:00:00+07","1970-01-19 00:00:00+07")
(4 rows)

DROP TABLE public.empl;
